zebrad/src/components/mempool/crawler.rs

```diff
@@ -57,43 +58,43 @@ where
     }
 
     /// Wait until the mempool is enabled.
-    async fn wait_until_enabled(&self) {
+    async fn wait_until_enabled(&mut self) {
         // TODO: Check if synchronizing up to chain tip has finished (#2603).
     }
 
     /// Crawl peers for transactions.
     ///
     /// Concurrently request [`FANOUT`] peers for transactions to include in the mempool.
-    async fn crawl_transactions(&self) {
+    async fn crawl_transactions(&mut self) {
         let requests = stream::repeat(Request::MempoolTransactionIds).take(FANOUT);
-        let peer_set = self.peer_set.lock().await.clone();
+        let peer_set = self.peer_set.clone();
 
         trace!("Crawling for mempool transactions");
 
         peer_set
             .call_all(requests)
             .unordered()
-            .and_then(|response| self.handle_response(response))
+            .and_then(handle_response)
             // TODO: Reduce the log level of the errors (#2655).
             .inspect_err(|error| info!("Failed to crawl peer for mempool transactions: {}", error))
             .for_each(|_| async {})
             .await;
     }
+}
 
-    /// Handle a peer's response to the crawler's request for transactions.
-    async fn handle_response(&self, response: Response) -> Result<(), BoxError> {
-        let transaction_ids = match response {
-            Response::TransactionIds(ids) => ids,
-            _ => unreachable!("Peer set did not respond with transaction IDs to mempool crawler"),
-        };
+/// Handle a peer's response to the crawler's request for transactions.
+async fn handle_response(response: Response) -> Result<(), BoxError> {
```

Comment:
We can't take `&mut self` here without cloning the crawler.
And generic type inference fails if it's a static method.

We'll probably just end up sending these responses to a channel anyway.

Reply:
> Hmm, I think I'm not a fan of having this method become a function. To me it seems easier to have it as a method because it will need to use the channel that's likely to be stored inside the `Crawler` type.

Ok, I've changed it to a static method in commit 509f1fcca4950c6069f762d3021999f12798ed98.
And added a stub channel, to make sure that will all work.

What do you think?

> Wouldn't it be simpler to just wait until we have updated the Tokio version, and see if that fixes the issue in a more straightforward manner?

I realise this question is a bit out of date - we're going to do the tokio upgrade after the mempool deadline.

I think having an instance method will work, but the ownership is a bit tricky to get right.
For an example, see PR #2661

We'll need to make sure all the fields are `Clone` and `Sync`.
But by the time we get around to that refactor, we'll have all the fields in place, so we'll know if that's reasonable.

Do you want to open a ticket for the post-tokio-upgrade refactor?